{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "initial_id",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-12-27T10:34:21.731596Z",
     "start_time": "2025-12-27T10:34:15.230586Z"
    },
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/caspar/thesis_code/complete_project/notebooks\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/caspar/miniconda3/envs/train_model_gpu/lib/python3.10/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'repr' attribute with value False was provided to the `Field()` function, which has no effect in the context it was used. 'repr' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.\n",
      "  warnings.warn(\n",
      "/home/caspar/miniconda3/envs/train_model_gpu/lib/python3.10/site-packages/pydantic/_internal/_generate_schema.py:2249: UnsupportedFieldAttributeWarning: The 'frozen' attribute with value True was provided to the `Field()` function, which has no effect in the context it was used. 'frozen' is field-specific metadata, and can only be attached to a model field using `Annotated` metadata or by assignment. This may have happened because an `Annotated` type alias using the `type` statement was used, or if the `Field()` function was attached to a single member of a union type.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import sys\n",
    "import os\n",
    "print(os.getcwd())\n",
    "os.chdir('../../complete_project/../')\n",
    "#\n",
    "# print(os.getcwd())\n",
    "\n",
    "# Then set up the paths\n",
    "# import sys\n",
    "\n",
    "\n",
    "# import os\n",
    "\n",
    "# os.environ['PYTHONPATH'] = os.getcwd()  # Now points to thesis_code directory\n",
    "# sys.path.append(os.environ['PYTHONPATH'])\n",
    "#you have to fix this yourself to import it correctly\n",
    "sys.path.append(\"/home/caspar/thesis_code/CellOracle\")\n",
    "sys.path.append(\"/home/caspar/thesis_code/complete_project/py files\")\n",
    "sys.path.append(\"/home/caspar/thesis_code/complete_project/py files/AIFiles\")\n",
    "sys.path.append(\"/home/caspar/thesis_code/complete_project/py files/AIFiles/batches\")\n",
    "sys.path.append(\"/home/caspar/thesis_code/complete_project/py files/AIFiles/single\")\n",
    "sys.path.append(\"/home/caspar/thesis_code/complete_project/py files/baseGRNConstructionFiles\")\n",
    "sys.path.append(\"/home/caspar/thesis_code/complete_project/py files/oracleInferenceFiles\")\n",
    "sys.path.append(\"/home/caspar/thesis_code/complete_project/py files/oracleSetup\")\n",
    "# import scanpy as sc\n",
    "# import pandas as pd\n",
    "# import numpy as np\n",
    "# import matplotlib.pyplot as plt\n",
    "import os\n",
    "# import anndata\n",
    "import sys\n",
    "import logging\n",
    "from datetime import datetime\n",
    "import wandb\n",
    "#import modified_celloracle as mco\n",
    "# from CellOracle import celloracle as co\n",
    "\n",
    "# import CellOracleSetup as setup_module\n",
    "# import GRNClusterAnalysis as analysis_module\n",
    "# import GRNInference as inference_module\n",
    "# import GRNInferenceTest as inference_test_module\n",
    "\n",
    "log_dir = 'logs'\n",
    "os.makedirs(log_dir, exist_ok=True)\n",
    "log_filename = os.path.join(log_dir, f\"app_{datetime.now().strftime('%Y_%m_%d')}.log\")\n",
    "\n",
    "# Configure the basic logging\n",
    "logging.basicConfig(\n",
    "    filename=log_filename,\n",
    "    filemode='a',\n",
    "    level=logging.INFO,\n",
    "    format='%(asctime)s - %(levelname)s - %(message)s'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "72f6ea408eeaed41",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-12-27T11:11:11.462573Z",
     "start_time": "2025-12-27T11:10:19.060079Z"
    }
   },
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'velocyto'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 4\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mscanpy\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01msc\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mCreateNewTFToTargetGeneList\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mtf_module\u001b[39;00m\n\u001b[0;32m----> 4\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mCellOracleSetupWithTFDict\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01msetup_module\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mCreateTransitionMatrix\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mtransition_module\u001b[39;00m\n",
      "File \u001b[0;32m~/thesis_code/complete_project/py files/oracleSetup/CellOracleSetupWithTFDict.py:9\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01msetuptools\u001b[39;00m\n\u001b[1;32m      8\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01msite\u001b[39;00m\n\u001b[0;32m----> 9\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mCellOracle\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m celloracle \u001b[38;5;28;01mas\u001b[39;00m co\n\u001b[1;32m     11\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mlogging\u001b[39;00m\n\u001b[1;32m     15\u001b[0m \u001b[38;5;124;03m\"\"\"This class is to setup celloracle with the base grn and train it on the scRNA-seq data\u001b[39;00m\n\u001b[1;32m     16\u001b[0m \u001b[38;5;124;03mit can load it in the base grn, but not the links! this is to just setup celloracle with the scRNA-seq data (or load in) and retrain the cluster speciric grns and save  to a file\u001b[39;00m\n\u001b[1;32m     17\u001b[0m \u001b[38;5;124;03minference and other stuff and loading in back links is done somewhere else\"\"\"\u001b[39;00m\n",
      "File \u001b[0;32m~/thesis_code/CellOracle/celloracle/__init__.py:8\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mwarnings\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mlogging\u001b[39;00m\n\u001b[0;32m----> 8\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m utility, network, network_analysis, go_analysis, data, data_conversion, oracle_utility\n\u001b[1;32m      9\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtrajectory\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01moracle_core\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m Oracle\n\u001b[1;32m     10\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mnetwork\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m Net\n",
      "File \u001b[0;32m~/thesis_code/CellOracle/celloracle/utility/__init__.py:18\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmake_log\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m makelog\n\u001b[1;32m      9\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mutility\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m (save_as_pickled_object, load_pickled_object,\n\u001b[1;32m     10\u001b[0m                       intersect,\n\u001b[1;32m     11\u001b[0m                       exec_process,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     16\u001b[0m                       knn_data_transferer,\n\u001b[1;32m     17\u001b[0m                       update_adata)\n\u001b[0;32m---> 18\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mload_hdf5\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m load_hdf5\n\u001b[1;32m     20\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mpandas_utility_for_jupyternotebook\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m init_datatable_mode\n\u001b[1;32m     22\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mpackage_version_checker\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m check_python_requirements\n",
      "File \u001b[0;32m~/thesis_code/CellOracle/celloracle/utility/load_hdf5.py:5\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# -*- coding: utf-8 -*-\u001b[39;00m\n\u001b[0;32m----> 5\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mnetwork_analysis\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m load_links\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mnetwork\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m load_net\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtrajectory\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01moracle_core\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m load_oracle\n",
      "File \u001b[0;32m~/thesis_code/CellOracle/celloracle/network_analysis/__init__.py:9\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m gene_analysis\n\u001b[1;32m      8\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m network_structure_analysis\n\u001b[0;32m----> 9\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mnetwork_construction\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m get_links\n\u001b[1;32m     10\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01muse_r_scripts\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m test_R_libraries_installation, get_R_path, set_R_path, config, _check_R_libraries_installation\n\u001b[1;32m     11\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mlinks_object\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m load_links, Links\n",
      "File \u001b[0;32m~/thesis_code/CellOracle/celloracle/network_analysis/network_construction.py:26\u001b[0m\n\u001b[1;32m     22\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mscipy\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m stats\n\u001b[1;32m     24\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mtqdm\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mauto\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m tqdm\n\u001b[0;32m---> 26\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mnetwork\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mnet_core\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m Net\n\u001b[1;32m     27\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mutility\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m standard\n\u001b[1;32m     28\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mlinks_object\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m Links\n",
      "File \u001b[0;32m~/thesis_code/CellOracle/celloracle/network/__init__.py:7\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;124;03mThe :mod:`.network` module implements GRN inference.\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \n\u001b[1;32m      5\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mnet_core\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m Net\n\u001b[0;32m----> 7\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mnet_util\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m load_net_from_patquets, getDF_TGxTF, getDF_peakxTF, load_net\n\u001b[1;32m     10\u001b[0m __all__ \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNet\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     11\u001b[0m            \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mload_net_from_patquets\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     12\u001b[0m            \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgetDF_TGxTF\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     13\u001b[0m            \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgetDF_peakxTF\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     14\u001b[0m            \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mload_net\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n",
      "File \u001b[0;32m~/thesis_code/CellOracle/celloracle/network/net_util.py:26\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mutility\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m load_pickled_object\n\u001b[1;32m     24\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mnet_core\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m Net\n\u001b[0;32m---> 26\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mvelocyto\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mserialization\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m dump_hdf5, load_hdf5\n\u001b[1;32m     28\u001b[0m \u001b[38;5;66;03m##############################\u001b[39;00m\n\u001b[1;32m     29\u001b[0m \u001b[38;5;66;03m### 1.  define main class  ###\u001b[39;00m\n\u001b[1;32m     30\u001b[0m \u001b[38;5;66;03m##############################\u001b[39;00m\n\u001b[1;32m     33\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mload_net\u001b[39m(file_path):\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'velocyto'"
     ]
    }
   ],
   "source": [
    "#make sure all packages are downloaded otherwise it will be quite annoying! packages were very diffiult to get to wor\n",
    "#especiallly cupy with the celloc version, i split the setup of celloracle and the running with inference of cupy into 2 different\n",
    "#cuda environemtns!\n",
    "\n",
    "import FilterCompleteData as filter_module\n",
    "import scanpy as sc\n",
    "import CreateNewTFToTargetGeneList as tf_module\n",
    "import CellOracleSetupWithTFDict as setup_module\n",
    "import CreateTransitionMatrix as transition_module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2008afba1ef378c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "FILTERED_AND_NORMALIZED =('filtered_and_normalized.h5ad', 'filtered_and_normalized', 'og_files' )\n",
    "PROMOTER_K4ME3_COUNTS = ('promoter_k4me3_counts.h5ad', 'promoter_k4me3_counts', 'og_files')\n",
    "RAW_COUNTS_FOR_PEARSON_FILTER = ('raw_counts_for_pearson_filter.h5ad', 'raw_counts_for_pearson_filter' , 'og_files')\n",
    "RAW_COUNTS_FOR_PEARSON_FILTER_SUBSET = ('raw_counts_for_pearson_filter_subset_from_normalized_set.h5ad', 'raw_counts_for_pearson_filter_subset_from_normalized_set','preprocessed_files_before_pearson_filter')\n",
    "\n",
    "#adjust this path as needed\n",
    "def load_in_data(file_name) -> sc.AnnData:\n",
    "    path = os.path.join(\"../../../celloracle_data\", file_name)\n",
    "    adata = sc.read_h5ad(path)\n",
    "    \n",
    "#this will error but you just have to load in your own data, more to show the flow of it all \n",
    "raw_data = load_in_data(RAW_COUNTS_FOR_PEARSON_FILTER[0])\n",
    "filtered_and_normalized_data = load_in_data(FILTERED_AND_NORMALIZED[0])\n",
    "\n",
    "#this is the path for the tf target gene list, this direclty creates the baseGRN, there are 3 other ways to adjust bae grn, we work with a tf to target gene list\n",
    "#if you want to do it from a atacseq dataset, lookup celloc documentation\n",
    "data_path = \"../celloracle_data/files_for_basegrn_creation\"\n",
    "input_tsv_path = os.path.join(data_path, \"tf_target_gene_list-4-28.tsv\")\n",
    "input_to_compare_old = os.path.join(data_path, \"promoter_to_tf_list.csv\")\n",
    "\n",
    "reduced_data = filter_module.FilterData(raw_data, filtered_and_normalized_data, \"mESCs\")\n",
    "tf_dict = tf_module.create_dict_from_csv(reduced_data)\n",
    "\n",
    "#this clal will setup the grn completely, base grn, cluster specific grns anf filtering and than saving the links! this will take a while to train!\n",
    "setup_obj = setup_module.CellOracleSetup(tf_dict, reduced_data, \"celltype\", \"X_umap\", \"output\", False, False, load_dir =\"\")\n",
    "\n",
    "#so now its all trained right in theory, we can use it to train the rl agent on or just compute the transition matrix, be aware that you use the modified and not the npm install default celloracle if you want to do targeted perturbations\n",
    "\n",
    "#depends if you want to do a transiton matrix or not, adjust it as you want, if its not already made, call this code\n",
    "transition_module.createTransitionMatrix(pahth_cellOc= \"somepath\", output_path=\"someoutputpath\", allow_OE=True)\n",
    "\n",
    "#now everything is ready for the rl agent to be trained on it, we will ignore batch as we build purely for the single env transiton matrix approach with possibilities of running parallel envs\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "896cd4013d8a84cb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-12-27T11:13:47.377172Z",
     "start_time": "2025-12-27T11:13:40.692531Z"
    }
   },
   "outputs": [],
   "source": [
    "#start the parameter tuning!\n",
    "import HPO_AULC_phase_2 as tuning_module2 # Import the refactored script\n",
    "import HPO_AULC_phase_1 as tuning_module # Import the refactored script\n",
    "\n",
    "\n",
    "#make sure you have the wandb login and project setup correctly as we monitor it, currently my project name is setup, but you have to login on your account! as on a different device this should not work (its free for academics), you have to adjust the rpoject name  in the singeenvrunbuildai.py file\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning, module=\"pandas\")\n",
    "\n",
    "\n",
    "env_type = \"NEW_INPUT\"\n",
    "data_path_new_data = os.path.join('../celloracle_data', \"celloracle_object/new_promoter_without_mescs_trimmed_test_own_umap\")\n",
    "save_path = os.path.join(\"../celloracle_data\", \"models\")\n",
    "notebook_config = {\n",
    "    # Paths\n",
    "    \"ORACLE_PATH\": os.path.join(data_path_new_data, \"ready_oracle.pkl\"),\n",
    "    \"MODEL_SAVE_PATH_BASE\": os.path.join(save_path, \"final_optuna_please\"), # Base for HPO trial models\n",
    "    \"MODEL_SAVE_PATH\": os.path.join(save_path, datetime.now().strftime('%Y_%m_%d_%H_%M_%S') + \"_model_\"+\"FINAL_MODEL_OPTIMIZED_PARAMS_NEW_INPUT\"), # Save model with timestamp\n",
    "    \"LOG_DIR_BASE\": os.path.join('../celloracle_data', \"final_optuna_please\"),         # Base for HPO trial logs\n",
    "    #set wand id to current timestamp\n",
    "    \"WANDB_RUN_ID\": datetime.now().strftime('%Y_%m_%d_%H_%M_%S'),\n",
    "    \"WAND_B_LOG_DIR_BASE\": os.path.join('../celloracle_data', \"final_optuna_please\"), # Base for HPO trial wandb logs\n",
    "    \"TRANSITION_MATRIX\": os.path.join(\"../celloracle_data/transition_matrix\", \"transition_matrix_int32.pkl\"), # Path to transition matrix\n",
    "    \"ORACLE_ADATA_PATH\": os.path.join(data_path_new_data, \"adata.pkl\"),\n",
    "    \"PERTURBABLE_GENES_PATH\": os.path.join(data_path_new_data, \"perturbable_genes.pkl\"),\n",
    "    \"RANDOM_SEED\" : 77,\n",
    "\n",
    "\n",
    "    # Training Setup\n",
    "    \"BATCH_SIZE\": 256,            # Smaller batch for potentially faster iteration in notebook\n",
    "    \"TOTAL_TIMESTEPS\": 100000,\n",
    "    \"TOTAL_TIMESTEPS_HPO\" : 2000000,\n",
    "    \"USE_MASKABLE_PPO\": False,    # Explicitly choose (make sure sb3_contrib is installed)\n",
    "    \"VERBOSE\": 0,\n",
    "    \"LOG_INTERVAL\": 10,\n",
    "    \"RESET_NUM_TIMESTEPS\": False,\n",
    "    \"DEVICE\": \"auto\",            # Use GPU if available\n",
    "    \"USE_PROGRESS_BAR\": True,    # Show progress in notebook output\n",
    "    \"STEP_SAVE_FREQ\":500000,\n",
    "\n",
    "    # CellOracle Parameters\n",
    "\n",
    "    # PPO Hyperparameters\n",
    "    \"PPO_N_STEPS\": 2048,      # Adjust PPO steps,  number of steps before ppo \"learns\"\n",
    "    \"PPO_N_EPOCHS\": 10,\n",
    "    \"PPO_BATCH_SIZE\": 128,\n",
    "    \"GAMMA\": 0.974,\n",
    "    \"GAE_LAMBDA\": 0.95,\n",
    "    \"CLIP_RANGE\":0.128,\n",
    "    \"ENT_COEF\": 0.00032,\n",
    "    \"VF_COEF\": 0.5,\n",
    "    \"MAX_GRAD_NORM\": 0.5,\n",
    "\n",
    "    # Environment Specific\n",
    "\n",
    "    \"ALLOW_GENE_ACTIVATION\": True,\n",
    "    \"STANDARD_SD_FACTOR\": 1.5,\n",
    "    \"USE_LEARNING_RATE_SCHEDULE\": True,\n",
    "    \"USE_SIMILARITY_REWARD\": True,\n",
    "    \"USE_PREV_KNOCKOUT\":True,\n",
    "    \"ENV_TYPE\": \"FAST\",  #options: FIRST,NEW_INPUT, GENERAL,FAST\n",
    "    \"ADD_NOISE\":False,\n",
    "    \"VAL_FRAC\":0.015,\n",
    "    \"TEST_FRAC\":0.015,\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    #ENV HYPERPARAMETERS TO TUNE\n",
    "    \"TARGET_CELLS_PER_PHASE\": 6,\n",
    "    \"MAX_STEPS_PER_EPISODE\": 20,\n",
    "    \"MAX_STEP_INCREASE_PER_PHASE\": 4,\n",
    "    \"MAX_STEPS_FIRST_PHASE\":150000,\n",
    "    \"PHASE_STEP_INCREASE\": 150000,\n",
    "    \"STEP_PENALTY\": -1,\n",
    "    \"GOAL_BONUS\": 0,\n",
    "    \"DISTANCE_REWARD_SCALE\": 5, #the higher the reward, the more we reward/punish getting closer\n",
    "    \"LEARNING_RATE\": 1.66e-4,\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    # NN Architecture\n",
    "    \"NET_WIDTH_FIRST_P\" : 256,\n",
    "    \"NET_WIDTH_SECOND_P\" : 256,\n",
    "    \"NET_WIDTH_FIRST_V\" : 256,\n",
    "    \"NET_WIDTH_SECOND_V\" : 64,\n",
    "    \"ACTIVATION_FN\": \"relu\",    # Specify activation\n",
    "    \"N_ENVS\" : 8,\n",
    "    \"HPO_MODE\" : True,\n",
    "    \"STUDY_NAME\": \"studyName_HPO_AULC_phase_1_with_date_\" + datetime.now().strftime('%Y_%m_%d_%H_%M_%S'),\n",
    "    \"STORAGE_NAME\": \"sqlite:///studyName_HPO_AULC_phase_1_with_date_\" + datetime.now().strftime('%Y_%m_%d_%H_%M_%S') + \".db\",\n",
    "    \n",
    "}\n",
    "if notebook_config[\"USE_SIMILARITY_REWARD\"]:\n",
    "    notebook_config[\"DISTANCE_REWARD_SCALE\"] =5\n",
    "else:\n",
    "\n",
    "    notebook_config[\"DISTANCE_REWARD_SCALE\"] = 5\n",
    "\n",
    "\n",
    "#so this does like 100 runs to tune params, you can adjust params in the HPO_AULC_phase_1.py file as needed\n",
    "trained_model = tuning_module.run_optuna_hpo(notebook_config)\n",
    "\n",
    "#now we have to do phase 2 of the tuning to finetune the parameters even more, it loads the top files from a save path\n",
    "\n",
    "notebook_config_phase_2 = notebook_config.copy()\n",
    "notebook_config_phase_2[\"TOP_PARAMS_PKL_PATH\"] = os.path.join(os.path.join(save_path, \"final_optuna_please\"), \"studyName_HPO_AULC_phase_1_with_date_2025_10_22_22_08_24_top_5_params.pkl\") #this is my path\n",
    "\n",
    "trained_model_phase_2 = tuning_module2.run_optuna_hpo_phase_2(notebook_config_phase_2)\n",
    "#these params of the phas 2 (so the best set) is saved to a file, now we use this as the main input\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4847a61180000353",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-12-27T11:20:04.300151Z",
     "start_time": "2025-12-27T11:19:57.650294Z"
    }
   },
   "outputs": [],
   "source": [
    "#from that phase 2 file now retrieve the params and insert in the dicoitnary config below!\n",
    "            ###FOR SINGLE INSTANCES!!!###\n",
    "import os\n",
    "import torch.nn as nn\n",
    "import SingleEnvRunBuildAI as single_ai_module # Import the refactored script\n",
    "import warnings\n",
    "\n",
    "# Add this at the top of your notebook cell before running training\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning, module=\"pandas\")\n",
    "\n",
    "env_type = \"NEW_INPUT\"\n",
    "data_path_new_data = os.path.join('../celloracle_data', \"celloracle_object/new_promoter_without_mescs_trimmed_test_own_umap\")\n",
    "save_path = os.path.join(\"../celloracle_data\", \"models\")\n",
    "notebook_config = {\n",
    "    # Paths\n",
    "    \"ORACLE_PATH\": os.path.join(data_path_new_data, \"ready_oracle.pkl\"),\n",
    "    \"MODEL_SAVE_PATH\": os.path.join(save_path, datetime.now().strftime('%Y_%m_%d_%H_%M_%S') + \"_model_\"+\"FINAL_MODEL_OPTIMIZED_PARAMS_NEW_INPUT\"), # Save model with timestamp\n",
    "    \"LOG_DIR\": os.path.join(data_path_new_data, \"logs_notebook_exp1\"), # Unique logs\n",
    "    #set wand id to current timestamp\n",
    "    \"WANDB_RUN_ID\": datetime.now().strftime('%Y_%m_%d_%H_%M_%S'),\n",
    "    \"WAND_B_LOG_DIR_BASE\": os.path.join('../celloracle_data', \"optuna_hpo_wandb_logs2\"), # Base for HPO trial wandb logs\n",
    "    \"TRANSITION_MATRIX\": os.path.join(\"../celloracle_data/transition_matrix\", \"transition_matrix_int32.pkl\"), # Path to transition matrix\n",
    "    \"ORACLE_ADATA_PATH\": os.path.join(data_path_new_data, \"adata.pkl\"),\n",
    "    \"PERTURBABLE_GENES_PATH\": os.path.join(data_path_new_data, \"perturbable_genes.pkl\"),\n",
    "    \"RANDOM_SEED\" : 77,\n",
    "    \"N_EVAL_EPISODES\": 1000,\n",
    "\n",
    "\n",
    "\n",
    "    # Training Setup\n",
    "\n",
    "\n",
    "    \"BATCH_SIZE\": 256,            # Smaller batch for potentially faster iteration in notebook\n",
    "    \"TOTAL_TIMESTEPS\": 2500000,\n",
    "    \"TOTAL_TIMESTEPS_HPO\" : 30000,\n",
    "    \"USE_MASKABLE_PPO\": False,    # Explicitly choose (make sure sb3_contrib is installed)\n",
    "    \"VERBOSE\": 0,\n",
    "    \"LOG_INTERVAL\": 10,\n",
    "    \"RESET_NUM_TIMESTEPS\": False,\n",
    "    \"DEVICE\": \"auto\",            # Use GPU if available\n",
    "    \"USE_PROGRESS_BAR\": True,    # Show progress in notebook output\n",
    "    \"STEP_SAVE_FREQ\":500000,\n",
    "\n",
    "    # CellOracle Parameters\n",
    "\n",
    "    # PPO Hyperparameters\n",
    "    \"PPO_N_STEPS\": 1024,      # Adjust PPO steps,  number of steps before ppo \"learns\"\n",
    "    \"PPO_N_EPOCHS\": 11,\n",
    "    \"PPO_BATCH_SIZE\": 64,\n",
    "    \"GAMMA\": 0.9831,\n",
    "    \"GAE_LAMBDA\": 0.9013,\n",
    "    \"CLIP_RANGE\":0.2141,\n",
    "    \"ENT_COEF\": 0.0054,\n",
    "    \"VF_COEF\": 0.4075,\n",
    "    \"MAX_GRAD_NORM\": 0.6452,\n",
    "\n",
    "    # Environment Specific\n",
    "\n",
    "    \"ALLOW_GENE_ACTIVATION\": True,\n",
    "    \"STANDARD_SD_FACTOR\": 1.5,\n",
    "    \"USE_LEARNING_RATE_SCHEDULE\": True,\n",
    "    \"ADD_NOISE\":False,\n",
    "    \"VAL_FRAC\":0.05,\n",
    "    \"TEST_FRAC\":0.05,\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    #ENV HYPERPARAMETERS TO TUNE\n",
    "    \"TARGET_CELLS_PER_PHASE\": 6,\n",
    "    \"MAX_STEPS_PER_EPISODE\": 20,\n",
    "    \"MAX_STEP_INCREASE_PER_PHASE\": 4,\n",
    "    \"MAX_STEPS_FIRST_PHASE\":100000,\n",
    "    \"PHASE_STEP_INCREASE\": 100000,\n",
    "    \"STEP_PENALTY\": -1,\n",
    "    \"GOAL_BONUS\": 0,\n",
    "    \"DISTANCE_REWARD_SCALE\": 5, #the higher the reward, the more we reward/punish getting closer\n",
    "    \"LEARNING_RATE\": 0.000107,\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    # NN Architecture\n",
    "    \"NET_WIDTH_FIRST_P\" : 256,\n",
    "    \"NET_WIDTH_SECOND_P\" : 64,\n",
    "    \"NET_WIDTH_FIRST_V\" : 128,\n",
    "    \"NET_WIDTH_SECOND_V\" : 64,\n",
    "    \"ACTIVATION_FN\": \"leaky_relu\",    # Specify activation\n",
    "    \"N_ENVS\" : 8,\n",
    "    \"HPO_MODE\" : False\n",
    "}\n",
    "    \n",
    "    #wat op dit moment nog niet werkt: the unique target types moet zijn waaruit we kunnen kiezen, niet op dit moment in de envs, dat is raar en nog niet deze: path_efficiency_ratio\n",
    "\n",
    "trained_model_main = single_ai_module.run_training(notebook_config)\n",
    "\n",
    "#sio this model is now trained and tested\n",
    "print(\"Training complete. Model saved at:\", notebook_config[\"MODEL_SAVE_PATH\"])\n",
    "\n",
    "#there are many files in the single thing to evaluate on the test set and generate reports and graphs etc\n",
    "#you should not run the eval with the mask, this is a mistake i made, in theory we should not even use a mask, this is owever the celloracle adjustments problem, with something like scGPT this would not be a problem at all\n",
    "#i am working on a fully bfs approach with ai agentic capabilities that tries to find valid paths by searching a vector database that matches transitions states to known transitons and thereby adjust possible paths, this just started and is a veyr parttime job so this will take a while but its an interesting approach i would say\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73089fe659a703e1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
